File: Java/백준/Silver/10866. 덱/덱.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.OutputStreamWriter;
import java.io.BufferedWriter;
import java.util.*;
import java.lang.Math;

public class Main {
    static InputStreamReader reader = new InputStreamReader(System.in);
    static BufferedReader br = new BufferedReader(reader);

    static OutputStreamWriter writer = new OutputStreamWriter(System.out);
    static BufferedWriter bw = new BufferedWriter(writer);

    public static void main(String[] args) throws IOException {
        int opCnt = Integer.parseInt(br.readLine());
        Deque dq = new Deque(10001);
        for (int i = 0; i < opCnt; i++) {
            String inp[] = br.readLine().split(" ");
            if (inp[0].equals("push_front")) {
                dq.pushFront(Integer.parseInt(inp[1]));
            }
            else if (inp[0].equals("push_back")) {
                dq.pushBack(Integer.parseInt(inp[1]));
            }
            else if (inp[0].equals("pop_front")) {
                bw.write(Integer.toString(dq.popFront()) + "\n");
            }
            else if (inp[0].equals("pop_back")) {
                bw.write(Integer.toString(dq.popBack()) + "\n");
            }
            else if (inp[0].equals("size")) {
                bw.write(Integer.toString(dq.size()) + "\n");
            }
            else if (inp[0].equals("empty")) {
                bw.write(Integer.toString(dq.isEmpty()) + "\n");
            }
            else if (inp[0].equals("front")) {
                bw.write(Integer.toString(dq.front()) + "\n");
            }
            else if (inp[0].equals("back")) {
                bw.write(Integer.toString(dq.back()) + "\n");
            }
        }
        bw.flush();
        bw.close();
    }
}
class Deque {
    private int container[];
    private int maxSize = 0;

    private int startIdx = 0;
    private int endIdx = 0;
    private int currentItemsCnt = 0;
    
    Deque(int size) {
        this.maxSize = size;
        this.container = new int[size];
    }

    void pushFront(int X) {
        if (this.isEmpty() == 1) {
            this.startIdx = 0; this.endIdx = 0;
            this.container[this.startIdx] = X;
            this.startIdx = (this.startIdx - 1 + this.maxSize) % this.maxSize;
            this.endIdx = (this.endIdx + 1 + this.maxSize) % this.maxSize;
        } else {
            this.container[this.startIdx] = X;
            this.startIdx = (this.startIdx - 1 + this.maxSize) % this.maxSize;
        }
        this.currentItemsCnt += 1;
    }

    void pushBack(int X) {
        if (this.isEmpty() == 1) {
            this.startIdx = 0; this.endIdx = 0;
            this.container[this.endIdx] = X;
            this.startIdx = (this.startIdx - 1 + this.maxSize) % this.maxSize;
            this.endIdx = (this.endIdx + 1 + this.maxSize) % this.maxSize;
        } else {
            this.container[this.endIdx] = X;
            this.endIdx = (this.endIdx + 1 + this.maxSize) % this.maxSize;
        }
        this.currentItemsCnt += 1;
    }

    int popFront() {
        if (this.isEmpty() == 1) return -1;
        this.currentItemsCnt -= 1;
        this.startIdx = (this.startIdx + 1 + this.maxSize) % this.maxSize;
        return this.container[this.startIdx];
    }

    int popBack() {
        if (this.isEmpty() == 1) return -1;
        this.currentItemsCnt -= 1;
        this.endIdx = (this.endIdx - 1 + this.maxSize) % this.maxSize;
        return this.container[this.endIdx];
    }

    int size() {
        return this.currentItemsCnt;
    }

    int isEmpty() {
        return (this.currentItemsCnt == 0)?1:0;
    }

    int front() {
        if (this.isEmpty() == 1) return -1;
        return this.container[((this.startIdx + 1 + this.maxSize) % this.maxSize)];
    }

    int back() {
        if (this.isEmpty() == 1) return -1;
        return this.container[((this.endIdx - 1 + this.maxSize) % this.maxSize)];
    }
}

File: Python/백준/Silver/10866. 덱/덱.py
import sys
class deque:
    def __init__(self):
        self.container = {}
        self.head = 0
        self.tail = 0
    def __len__(self):
        return len(self.container)
    def append(self, n):
        self.container[self.tail] = n
        self.tail += 1
    def appendleft(self, n):
        self.head -= 1
        self.container[self.head] = n
    def pop(self):
        if len(self):
            self.tail -= 1
            tmp = self.container[self.tail]
            del self.container[self.tail]
            return tmp
        else:
            return -1
    def popleft(self):
        if len(self):
            tmp = self.container[self.head]
            del self.container[self.head]
            self.head += 1
            return tmp
        else:
            return -1
    def value(self, n):
        if len(self):
            return self.container[self.head + n]
        else:
            return -1

dq = deque()
input = sys.stdin.readline
N = int(input())
for _ in range(N):
    inp = list(map(str, (input().rstrip() + ' 0').split()))
    if inp[0] == 'push_front':
        dq.appendleft(inp[1])
    elif inp[0] == 'push_back':
        dq.append(inp[1])
    elif inp[0] == 'pop_front':
        print(dq.popleft())
    elif inp[0] == 'pop_back':
        print(dq.pop())
    elif inp[0] == 'empty':
        print(int(not bool(len(dq))))
    elif inp[0] == 'size':
        print(len(dq))
    elif inp[0] == 'front':
        print(dq.value(0))
    elif inp[0] == 'back':
        print(dq.value(len(dq) - 1))